diff --git a/mrbgem.rake b/mrbgem.rake
index 8120f78..51522e9 100644
--- a/mrbgem.rake
+++ b/mrbgem.rake
@@ -6,9 +6,9 @@ MRuby::Gem::Specification.new('mruby-io') do |spec|
   
   case RUBY_PLATFORM
   when /mingw|mswin/
-    spec.linker.libraries += ['Ws2_32']
+    spec.linker.libraries += ['ws2_32']
     #spec.cc.include_paths += ["C:/Windows/system/include"]
-    spec.linker.library_paths += ["C:/Windows/system"]
+    spec.linker.library_paths += ["C:/Windows/System"]
   end
   if build.kind_of?(MRuby::CrossBuild) && %w(x86_64-w64-mingw32 i686-w64-mingw32).include?(build.host_target)
     spec.linker.libraries += ['ws2_32']
